pkg/jobmanager/status.go

```diff
@@ -38,131 +46,183 @@ var TargetRoutingEvents = []event.Name{
 	target.EventTargetInErr,
 }
 
+// RunCoordinates collects information to identify the run for which we want to rebuild the status
+type RunCoordinates struct {
+	JobID types.JobID
+	RunID types.RunID
+}
+
+// TestCoordinates collects information to identify the test for which we want to rebuild the status
+type TestCoordinates struct {
+	RunCoordinates
+	TestName string
+}
+
+// TestStepCoordinates collects information to identify the test step for which we want to rebuild the status
+type TestStepCoordinates struct {
+	TestCoordinates
+	TestStepName  string
+	TestStepLabel string
+}
+
 // buildTargetStatus populates a TestStepStatus object with TestStepStatus information
-func (jm *JobManager) buildTargetStatus(jobID types.JobID, currentTestStepStatus *job.TestStepStatus) error {
+func (jm *JobManager) buildTargetStatuses(coordinates TestStepCoordinates) ([]job.TargetStatus, error) {
 
 	// Fetch all the events associated to Targets routing
 	routingEvents, err := jm.testEvManager.Fetch(
 		[]testevent.QueryField{
-			testevent.QueryJobID(jobID),
+			testevent.QueryJobID(coordinates.JobID),
+			testevent.QueryTestName(coordinates.TestName),
+			testevent.QueryTestStepLabel(coordinates.TestStepLabel),
 			testevent.QueryEventNames(TargetRoutingEvents),
 		},
 	)
 	if err != nil {
-		return fmt.Errorf("could not fetch events associated to target routing: %v", err)
+		return nil, fmt.Errorf("could not fetch events associated to target routing: %v", err)
 	}
 
-	// Filter only the events emitted by this TestStep, group events by Target
+	targetStatuses := make([]job.TargetStatus, 0)
 	for _, testEvent := range routingEvents {
-		if testEvent.Header.TestStepLabel != currentTestStepStatus.TestStepLabel {
-			continue
-		}
-		// Update the TargetStatus object associated to the Target. If there is no
-		// TargetStatus associated yet, append it
-		var currentTargetStatus *job.TargetStatus
-		for index, candidateTargetStatus := range currentTestStepStatus.TargetStatus {
-			if *candidateTargetStatus.Target == *testEvent.Data.Target {
-				currentTargetStatus = &currentTestStepStatus.TargetStatus[index]
+		// Update the TargetStatus object associated to the Target.
+		// If there is no TargetStatus associated yet, append it
+		var targetStatus *job.TargetStatus
+		for index, candidateStatus := range targetStatuses {
```

Comment:
Very subjective, just ignore it :) :

I think it would've been better to sort `routingEvents` by `Target` instead of doing a linear search on each step. Also I would move this whole `for` to a separate function (like `(job.TargetStatuses).ToTargetStatuses()`).